# Типы и модель данных
В данном уроке разберем как Python работает с переменными и определим какие типы данных можно использовать в рамках этого языка. Подробно рассмотрим модель данных Python, а также механизмы создания и изменения значения переменных.

## Немного о типизации языков программирования
Если достаточно формально подходить к вопросу о типизации языка Python, то можно сказать, что он относится к языкам с <b>неявной сильной динамической типизацией</b> (более подробная инофрмация доступна в <b><a href="https://www.dropbox.com/s/j5k0pxzrwbjymzt/%D0%9B%D0%B8%D0%BA%D0%B1%D0%B5%D0%B7%20%D0%BF%D0%BE%20%D1%82%D0%B8%D0%BF%D0%B8%D0%B7%D0%B0%D1%86%D0%B8%D0%B8%20%D0%B2%20%D1%8F%D0%B7%D1%8B%D0%BA%D0%B0%D1%85%20%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F.pdf?dl=0">PDF</a></b> или на <b><a href="https://habr.com/ru/post/161205/">Habr</a></b>).

### Неявная?
Явная / неявная типизация. Явно-типизированные языки отличаются тем, что тип новых переменных / функций / их аргументов нужно задавать явно. Соответственно языки с неявной типизацией перекладывают эту задачу на компилятор / интерпретатор.

Примеры:

Явная: C++, D, C#

Неявная: PHP, Lua, JavaScript

> При неявной типизации <b>задавать тип</b> переменных, функций и их аргументов <b>не нужно</b>. 

### Сильная?
Сильная / слабая типизация (также иногда говорят строгая / нестрогая). Сильная типизация выделяется тем, что язык не позволяет смешивать в выражениях различные типы и не выполняет автоматические неявные преобразования, например нельзя вычесть из строки множество. Языки со слабой типизацией выполняют множество неявных преобразований автоматически, даже если может произойти потеря точности или преобразование неоднозначно.

Примеры:

Сильная: Java, Python, Haskell, Lisp;

Слабая: C, JavaScript, Visual Basic, PHP.

> При сильной(строгой) типизации язык <b>не позволяет смешивать в выражениях различные типы и не выполняет автоматические неявные преобразования (преведение типов)</b>


### Динамическая?
Статическая / динамическая типизация. Статическая определяется тем, что конечные типы переменных и функций устанавливаются на этапе компиляции. Т.е. уже компилятор на 100% уверен, какой тип где находится. В динамической типизации все типы выясняются уже во время выполнения программы.

Примеры:

Статическая: C, Java, C#;

Динамическая: Python, JavaScript, Ruby.

> В динамической типизации <b>все типы выясняются уже во время выполнения программы</b>.

## Типы данных в Python
В Python типы данных можно разделить на встроенные в интерпретатор (built-in) и не встроенные, которые можно использовать при импортировании соответствующих модулей.

>К основным встроенным типам относятся:
- None (неопределенное значение переменной)
- Логические переменные (Boolean Type)
- Числа (Numeric Type):
        - int – целое число
        - float – число с плавающей точкой
        - complex – комплексное число
- Списки (Sequence Type)
        - list – список
        - tuple – кортеж
        - range – диапазон
- Строки (Text Sequence Type )
        - str
- Бинарные списки (Binary Sequence Types)
        - bytes – байты
        - bytearray – массивы байт
        - memoryview – специальные объекты для доступа к внутренним данным объекта через protocol buffer
- Множества (Set Types)
        - set – множество
        - frozenset – неизменяемое множество
- Словари (Mapping Types)
        - dict – словарь

## Модель данных
Рассмотрим как создаются объекты в памяти, их устройство, процесс объявления новых переменных и работу операции присваивания.

Для того, чтобы объявить и сразу инициализировать переменную необходимо написать её имя, потом поставить знак равенства и значение, с которым эта переменная будет создана. Например строка:

> b = 5

объявляет переменную b и присваивает ей значение 5.

Целочисленное значение 5 в рамках языка Python по сути своей является объектом. Объект, в данном случае – это абстракция для представления данных, данные – это числа, списки, строки и т.п. При этом, под данными следует понимать как непосредственно сами объекты, так и отношения между ними (об этом чуть позже). Каждый объект имеет три атрибута – это идентификатор, значение и тип. Идентификатор – это уникальный признак объекта, позволяющий отличать объекты друг от друга, а значение – непосредственно информация, хранящаяся в памяти, которой управляет интерпретатор

При инициализации переменной, на уровне интерпретатора, происходит следующее:

- создается целочисленный объект 5 (можно представить, что в этот момент создается ячейка и 5 кладется в эту ячейку);
- данный объект имеет некоторый идентификатор, значение: 5, и тип: целое число;
- посредством оператора “=” создается ссылка между переменной b и целочисленным объектом 5 (переменная b ссылается на объект 5).

Имя переменной не должно совпадать с ключевыми словами интерпретатора Python. Список ключевых слов можно найти <a href="https://pythonworld.ru/osnovy/klyuchevye-slova-modul-keyword.html">здесь</a>. Также его можно получить непосредственно в программе, для этого нужно подключить модуль <b>keyword</b> и воспользоваться командой <b>keyword.kwlist</b>.

In [1]:
import keyword
print("Python keywords: ", keyword.kwlist)

Python keywords:  ['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


Проверить является или нет идентификатор ключевым словом можно так:    

In [4]:
print(keyword.iskeyword("try"))
print(keyword.iskeyword("b"))

True
False


Для того, чтобы посмотреть на объект с каким идентификатором ссылается данная переменная, можно использовать функцию <b>id()</b>.

In [18]:
a = 4
b = 5
print("Идентификатор a =", id(a))
print("Идентификатор b =",id(b))
a = b
print("Идентификатор a после присвоения к b =",id(a))

Идентификатор a = 140727456735728
Идентификатор b = 140727456735760
Идентификатор a после присвоения к b = 140727456735760


Как видно из примера, идентификатор – это некоторое целочисленное значение, посредством которого уникально адресуется объект. Изначально переменная a ссылается на объект 4 с идентификатором 140727456735728, переменная b – на объект с id = 140727456735760. После выполнения операции присваивания a = b, переменная a стала ссылаться на тот же объект, что и b.

<img src="./source/images/example-1.png">

Тип переменной можно определить с помощью функции <b>type()</b>. Пример использования приведен ниже.

In [24]:
a = 10
b = "hello"
c = (1, 2)
print("a:", type(a))
print("b:", type(b))
print("c:", type(c))

a: <class 'int'>
b: <class 'str'>
c: <class 'tuple'>


## Изменяемые и неизменяемые типы данных

В Python существуют изменяемые и неизменяемые типы.

К <b>неизменяемым (immutable)</b> типам относятся: 
- целые числа (int)  
- числа с плавающей точкой (float) 
- комплексные числа (complex) 
- логические переменные (bool) 
- кортежи (tuple, 
- строки (str) 
- неизменяемые множества (frozen set)

К <b>изменяемым (mutable)</b> типам относятся: 
- списки (list) 
- множества (set) 
- словари (dict)

Как уже было сказано ранее, при создании переменной, вначале создается объект, который имеет уникальный идентификатор, тип и значение, после этого переменная может ссылаться на созданный объект.

Неизменяемость типа данных означает, что созданный объект больше не изменяется. Например, если мы объявим переменную k = 15, то будет создан объект со значением 15, типа int и идентификатором, который можно узнать с помощью функции id().

In [29]:
k = 15
print("Идентификатор к, адрес объекта на который ссылается к =",id(k))
print("Тип объекта на который ссылкается к, то есть 15 =",type(k))

Идентификатор к, адрес объекта на который ссылается к= 140727456736080
Тип объекта на который ссылкается к, то есть 15 = <class 'int'>


Объект с id = 140727456736080 будет иметь значение 15 и изменить его уже нельзя.

Если тип данных изменяемый, то можно менять значение объекта. Например, создадим список [1, 2], а потом заменим второй элемент на 3.

In [33]:
a = [1, 2]
print(a)
print("Идентификатор a =",id(a))
a[1] = 3
print(a)
print("Идентификатор a =",id(a))

[1, 2]
Идентификатор a = 1850587170056
[1, 3]
Идентификатор a = 1850587170056


Как видно, объект на который ссылается переменная a, был изменен. Это можно проиллюстрировать следующим рисунком.

<img src="./source/images/example-2.png">

В рассмотренном случае, в качестве данных списка, выступают не объекты, а отношения между объектами. Т.е. в переменной a хранятся ссылки на объекты содержащие числа 1 и 3, а не непосредственно сами эти числа.